File: songHistory.js
import { readFileSync, writeFileSync, existsSync } from 'fs';

const HISTORY_FILE = 'downloaded_songs.json';

function loadHistory() {
    try {
        if (existsSync(HISTORY_FILE)) {
            const data = readFileSync(HISTORY_FILE, 'utf8');
            return JSON.parse(data);
        }
    } catch (e) {
        console.log(`[SongHistory] Error loading history: ${e.message}`);
    }
    return { songs: [] };
}

function saveHistory(history) {
    try {
        writeFileSync(HISTORY_FILE, JSON.stringify(history, null, 2));
    } catch (e) {
        console.log(`[SongHistory] Error saving history: ${e.message}`);
    }
}

export function isAlreadyDownloaded(identifier) {
    const history = loadHistory();
    return history.songs.some(song => 
        song.url === identifier || 
        song.videoId === identifier ||
        song.title?.toLowerCase() === identifier?.toLowerCase()
    );
}

export function addToHistory(songInfo) {
    const history = loadHistory();
    
    const exists = history.songs.some(song => 
        song.url === songInfo.url || 
        song.videoId === songInfo.videoId
    );
    
    if (!exists) {
        history.songs.push({
            title: songInfo.title,
            url: songInfo.url,
            videoId: songInfo.videoId,
            downloadedAt: new Date().toISOString()
        });
        saveHistory(history);
        console.log(`[SongHistory] Added: ${songInfo.title}`);
        return true;
    }
    
    console.log(`[SongHistory] Already exists: ${songInfo.title}`);
    return false;
}

export function getDownloadedCount() {
    const history = loadHistory();
    return history.songs.length;
}

File: songHistory.py
import json
import os

HISTORY_FILE = 'downloaded_songs.json'

def load_history():
    try:
        if os.path.exists(HISTORY_FILE):
            with open(HISTORY_FILE, 'r') as f:
                return json.load(f)
    except Exception as e:
        print(f"[SongHistory] Error loading history: {e}")
    return {"songs": []}

def save_history(history):
    try:
        with open(HISTORY_FILE, 'w') as f:
            json.dump(history, f, indent=2)
    except Exception as e:
        print(f"[SongHistory] Error saving history: {e}")

def is_already_downloaded(identifier):
    history = load_history()
    for song in history['songs']:
        if (song.get('url') == identifier or 
            song.get('video_id') == identifier or
            (song.get('title') and identifier and 
             song.get('title', '').lower() == identifier.lower())):
            return True
    return False

def add_to_history(song_info):
    history = load_history()
    
    for song in history['songs']:
        if (song.get('url') == song_info.get('url') or 
            song.get('video_id') == song_info.get('video_id')):
            print(f"[SongHistory] Already exists: {song_info.get('title')}")
            return False
    
    from datetime import datetime
    history['songs'].append({
        'title': song_info.get('title'),
        'url': song_info.get('url'),
        'video_id': song_info.get('video_id'),
        'downloaded_at': datetime.now().isoformat()
    })
    save_history(history)
    print(f"[SongHistory] Added: {song_info.get('title')}")
    return True

def get_downloaded_count():
    history = load_history()
    return len(history['songs'])
